src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -1386,9 +1386,10 @@ public void CopyMetadataTo(ITaskItem destinationItem, bool addOriginalItemSpec)
             /// <returns>The cloned metadata.</returns>
             public IDictionary CloneCustomMetadata()
             {
-                Dictionary<string, string> clonedMetadata = new Dictionary<string, string>(MSBuildNameIgnoreCaseComparer.Default);
+                var metadata = MetadataCollection;
+                Dictionary<string, string> clonedMetadata = new Dictionary<string, string>(metadata.Count, MSBuildNameIgnoreCaseComparer.Default);
```

Comment:
Any reason you couldn't just use MetadataCollection.Count instead of making a new variable?

Reply:
Accessing MetadataCollection is expensive as it creates a new copy every time.

Reply:
This whole area has too many allocations. Hope one day we can switch to immutable collections and clean all of this up.

Reply:
Ah, I see. And you can't just cache it and avoid that problem because metadata can change, just not between these two points. Sounds good!